### ISSUE TITLE
not able to use USB devices on A10 Hackberry

### ISSUE THREAD

**Issue by tisdall** [linux-sunxi/linux-sunxi#158]
I tried upgrading from kernel 3.4.43 to 3.4.67 but now I'm not able to use any USB devices connected to the device.

I initially tried with sun4i_defconfig and added only btusb (Bluetooth USB support) but that didn't work.

I've tried various combinations of shutting on/off "USB Gadget Support" and setting the "USB0 Controller support" to each of the 4 options.

I found the followng in dmesg:

```
[    1.480399] sunxi_emac Using mii phy on PortA
[    1.499496] sunxi_emac Using MAC from SID: XX:XX:XX:XX:XX:XXERR: script_parser_fetch usb_controller_type failed
```

`lsusb` shows hub devices, but never anything else

**Comment by tisdall** [comment]
here are some other people with the same issue:  https://www.miniand.com/forums/forums/development--3/topics/custom-kernel-powering-usb-devices

One person mentioned that the issue was powering the USB bus and an externally powered hub was enough to fix the problem. (I'm not able to confirm that)

**Comment by tisdall** [comment]
Okay, just remembered a powered hub I have and can confirm that everything is working fine if I plug in a powered hub.  Anyone have any thoughts on what I could try to get the internal hub powered?

**Comment by arokux** [comment]
Hi,

it is known that this issue is not present in sunxi-v3.4.43-r1 (please
verify this is the case for you, before you continue) tag of linux-sunxi
and appears in sunxi-v3.4.43-r2. Unfortunately we should fallback to git
bisect, since I have no clue
what could be wrong (I have tried several things without success). This
means a bit of testing, but it is not
difficult. You will need to test about 7 kernels and git will assist
you at it.

Please read this small explanation about how git bisect works

http://git-scm.com/book/en/Git-Tools-Debugging-with-Git#Binary-Search

(I assume you have cloned the linux-sunxi kernel. Change to the linux-sunxi
directory and continue with further steps.)

Then proceed as follows:

git bisect start
git bisect bad sunxi-v3.4.43-r2
git bisect good sunxi-v3.4.43-r1

git will checkout the needed kernel for you, you'd just need to build
and boot it. Then tell git if it was good or bad by either

git bisect good

or

git bisect bad

Once you finished testing and know which commit introduces regression
perform

git bisect reset

I hope you can do this testing since I'm very curious now what is the
problem and I cannot reproduce the bug with my hardware.

Best regards,
Arokux

**Comment by romanrm** [comment]
Where did you get the script.bin that you use, and generally what is the device model?
Try using the script.bin that's stored on the stock /dev/nanda partition.

**Comment by tisdall** [comment]
@romanrm things were working fine with 3.4.43+ with the same script.bin so can't it be assumed that there's no issue with it?

The hardware is a Hackberry A10: https://www.miniand.com/products/Hackberry%20A10%20Developer%20Board  The message forum I linked to before is the message forum for that device.

**Comment by tisdall** [comment]
@arokux I'll see what I can do.  Hopefully it won't take me too long.

**Comment by arokux** [comment]
There is update from balage (IRC username), he has also commented in the
forum thread you have shown. The regression was introduced by one of the
commits

1d524c5e244b79a5a9570e028d2497fb17ffdfbb

8242195a7bb3d8ac3a829b8e7e0f226bd998d78a

Can you please confirm and identify by which exactly? You could use git
revert.

This is somehow strange since these commits are not touching usb host, but
usb otg. On the Internet I was able to find that the lower port on the
Hackberry is usb otg. Can you confirm that the upper part also lacks power?

Thanks

**Comment by tisdall** [comment]
@arokux Well, I'm already part way through the bisect thing so I might as well complete it. :)  

I have 2 dongles so I've plugged them both in and will continue testing for both.  I have noticed that some revisions show different numbers of "Linux Foundation X.X root hub" when running `lsusb`.  Is that significant?  I'm not really sure what those entries represent.

**Comment by romanrm** [comment]
> things were working fine with 3.4.43+ with the same script.bin so can't it be assumed that there's no issue with it?

No, not really.  Will not insist this is the case here, but it is possible the older kernel was "less strict" or did not use some values from script.bin (or did not check errors parsing them), and was  coincidentally setting up a working configuration despite an incorrect script.bin.

> This is somehow strange since these commits are not touching usb host, but
> usb otg.

Well at least the first one is the one actually adding the error message we see here in the bug report.

**Comment by techn** [comment]
Are we talking about USB0(aka. otg) or USB1/2 ?

https://github.com/linux-sunxi/linux-sunxi/commit/1d524c5e244b79a5a9570e028d2497fb17ffdfbb#diff-d4c071461b25da5462c18cae9b957a77R109

**Comment by tisdall** [comment]
@arokux I couldn't get 1d524c5e244b79a5a9570e028d2497fb17ffdfbb to even compile.  I got errors on drivers/usb/sunxi_usb/hcd/hcd0/swhcd0.c when compiling the uImage

**Comment by tisdall** [comment]
@romanrm checked the script.bin and it is different than the one on the nand.  I'm not that familiar with script.bin but the one I'm using is from a disk image someone created here: http://dl.miniand.com/dinky/hackbuntu-1gbRam-2gb.img.tar.gz  (if I'm remembering correctly)

**Comment by tisdall** [comment]
@techn all I know is it's not working  :)  I'm using the sun4i_defconfig except I'm also adding btusb.  If I use a powered external hub then it seems to work okay.  The USB device I'm using is just a Bluetooth dongle.

**Comment by techn** [comment]
@tisdall hackberry doesn't seem to have USB0, so only affecting rows in that patch are : https://github.com/linux-sunxi/linux-sunxi/commit/1d524c5e244b79a5a9570e028d2497fb17ffdfbb#diff-d4c071461b25da5462c18cae9b957a77R121 and https://github.com/linux-sunxi/linux-sunxi/commit/1d524c5e244b79a5a9570e028d2497fb17ffdfbb#diff-d4c071461b25da5462c18cae9b957a77R125 , so you might want to comment them.

Other should affect only on USB0.

**Comment by tisdall** [comment]
@arokux couldn't get 8242195a7bb3d8ac3a829b8e7e0f226bd998d78a to compile either with what looks like the same errors...

```
drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.c: In function ‘sw_usb_host0_enable’:
drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.c:1539:9: error: ‘g_hcd0_pdev’ undeclared (first use in this function)
drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.c:1539:9: note: each undeclared identifier is reported only once for each function it appears in
drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.c: In function ‘sw_usb_host0_disable’:
drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.c:1612:9: error: ‘g_hcd0_pdev’ undeclared (first use in this function)
drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.c: In function ‘sw_hcd_probe_otg’:
drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.c:1696:2: error: ‘g_hcd0_pdev’ undeclared (first use in this function)
drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.c: In function ‘sw_hcd_remove_otg’:
drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.c:1758:2: error: ‘g_hcd0_pdev’ undeclared (first use in this function)
drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.c: At top level:
drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.c:1684:12: warning: ‘sw_hcd_probe_otg’ defined but not used [-Wunused-function]
drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.c:1743:12: warning: ‘sw_hcd_remove_otg’ defined but not used [-Wunused-function]
make[2]: *** [drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [drivers/usb/sunxi_usb] Error 2
make: *** [drivers] Error 2
```

**Comment by techn** [comment]
@tisdall the is some config changes done/fixed after. You need to make sure that kernel config is correct

**Comment by tisdall** [comment]
@arokux result of `git bisect`:

```
8242195a7bb3d8ac3a829b8e7e0f226bd998d78a is the first bad commit
commit 8242195a7bb3d8ac3a829b8e7e0f226bd998d78a
Author: Jari Helaakoski <tekkuli@gmail.com>
Date:   Fri Jun 28 00:31:46 2013 +0300

    usb:sunxi:Import changes from Allwinner SDK. Minor bits.

    Compared to latest SDK. Minor bits. Shouldn't have any functional effect

    Signed-off-by: Jari Helaakoski <tekkuli@gmail.com>
    Signed-off-by: Hans de Goede <hdegoede@redhat.com>

:040000 040000 723931448c96b238a8bc72bbaf84b1a39f0bfaf5 3a9c2aad2bc818ccd313c05725ed8b6003436f71 M    drivers
```

Unfortunately this is mostly because I had to mark the uncompilable revisions as "bad".

**Comment by tisdall** [comment]
@techn sorry, what changes do I need to make in the kernel config to get that compiling?

**Comment by ssvb** [comment]
@tisdall 

> Unfortunately this is mostly because I had to mark the uncompilable revisions as "bad".

Normally it is better to "skip" uncompilable revisions. But if there are too many of them, git may be unable to identify a single problematic commit and just report a range of commits as the result of bisecting.

**Comment by tisdall** [comment]
@ssvb good to know for next time...  This was my first time using bisect so I had no idea "skip" was even an option.  ;)

**Comment by tisdall** [comment]
@techn from looking around, it seems that the hackberry does have an OTG port.  You use it to flash the internal NAND.

I'm going to take a look at the script.bin and see what the USB settings are in there.

**Comment by tisdall** [comment]
@techn okay, so the script.bin (both the one I'm using and the original on the NAND) are set to make the USB "host only" (which doesn't make a lot of sense because I'm pretty sure you use one port to flash the NAND).  Regardless, I tried compiling the latest revision in the sunxi-3.4 branch with the configuration set to "host only" and I still wasn't able to get any power on the device.

**Comment by techn** [comment]
@tisdall Retested with A10 mini-x and latest codes.. USB1/2 works ok, without hub.
And "[   10.111110] ERR: script_parser_fetch usb_controller_type failed " is printed here also.

**Comment by tisdall** [comment]
noticed this is `dmesg`...  statements about "Set USB Power ON" and "Set USB Power OFF"

```
[    1.477183] sunxi_emac Using mii phy on PortA
[    1.496297] sunxi_emac Using MAC from SID: 02:48:07:42:9f:e9
[    1.509193] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    1.521107] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    1.529625] [sw-ehci1]: open clock
[    1.555715] [sw-ehci1]: Set USB Power ON
[    1.566064] sw-ehci sw-ehci.1: SW USB2.0 'Enhanced' Host Controller (EHCI) Driver
[    1.579571] sw-ehci sw-ehci.1: new USB bus registered, assigned bus number 2
[    1.590995] sw-ehci sw-ehci.1: irq 39, io mem 0x01c14000
[    1.614832] sw-ehci sw-ehci.1: USB 2.0 started, EHCI 1.00
[    1.622842] ehci_irq: port change detect
[    1.630067] hub 2-0:1.0: USB hub found
[    1.636745] hub 2-0:1.0: 1 port detected
[    1.643166] [sw-ohci1]: open clock
[    1.672714] sw-ohci sw-ohci.1: SW USB2.0 'Open' Host Controller (OHCI) Driver
[    1.685872] sw-ohci sw-ohci.1: new USB bus registered, assigned bus number 3
[    1.697220] sw-ohci sw-ohci.1: irq 64, io mem 0x01c14400
[    1.767612] hub 3-0:1.0: USB hub found
[    1.774285] hub 3-0:1.0: 1 port detected
[    1.780652] [sw-ehci2]: open clock
[    1.806740] [sw-ehci2]: Set USB Power ON
[    1.817107] sw-ehci sw-ehci.2: SW USB2.0 'Enhanced' Host Controller (EHCI) Driver
[    1.830618] sw-ehci sw-ehci.2: new USB bus registered, assigned bus number 4
[    1.842035] sw-ehci sw-ehci.2: irq 40, io mem 0x01c1c000
[    1.965868] usb 2-1: new high-speed USB device number 2 using sw-ehci
[    1.994825] sw-ehci sw-ehci.2: USB 2.0 started, EHCI 1.00
[    2.002841] ehci_irq: port change detect
[    2.009959] hub 4-0:1.0: USB hub found
[    2.016635] hub 4-0:1.0: 1 port detected
[    2.023377] [sw-ehci2]: disable ehci
[    2.030343] sw-ehci sw-ehci.2: remove, state 1
[    2.038819] usb usb4: USB disconnect, device number 1
[    2.048500] sw-ehci sw-ehci.2: USB bus 4 deregistered
[    2.056296] [sw-ehci2]: Set USB Power OFF
[    2.062480] [sw-ehci2]: close clock
[    2.068273] [sw-ohci2]: open clock
[    2.094361] [sw-ohci2]: Set USB Power ON
[    2.104361] sw-ohci sw-ohci.2: SW USB2.0 'Open' Host Controller (OHCI) Driver
[    2.117519] sw-ohci sw-ohci.2: new USB bus registered, assigned bus number 4
[    2.128912] sw-ohci sw-ohci.2: irq 65, io mem 0x01c1c400
[    2.165429] hub 2-1:1.0: USB hub found
[    2.172451] hub 2-1:1.0: 4 ports detected
[    2.197632] hub 4-0:1.0: USB hub found
[    2.204303] hub 4-0:1.0: 1 port detected
[    2.211023] [sw-ohci2]: disable ohci
[    2.218023] sw-ohci sw-ohci.2: remove, state 1
[    2.226488] usb usb4: USB disconnect, device number 1
[    2.235997] sw-ohci sw-ohci.2: USB bus 4 deregistered
[    2.243790] [sw-ohci2]: Set USB Power OFF
[    2.249974] [sw-ohci2]: close clock
[    2.257350] Initializing USB Mass Storage driver...
[    2.267350] usbcore: registered new interface driver usb-storage
[    2.276993] USB Mass Storage support registered.
[    2.286600] usbcore: registered new interface driver ums-alauda
[    2.297582] usbcore: registered new interface driver ums-cypress
```

```
[    7.468498] ERR: script_parser_fetch usb_controller_type failed
[    7.470966] [sw-ehci2]: enable ehci
[    7.577603] [sw-ehci2]: open clock
[    7.920800] [sw-ehci2]: Set USB Power ON
[    7.980321] sw-ehci sw-ehci.2: SW USB2.0 'Enhanced' Host Controller (EHCI) Driver
[    8.056225] sw-ehci sw-ehci.2: new USB bus registered, assigned bus number 4
[    8.109703] sw-ehci sw-ehci.2: irq 40, io mem 0x01c1c000
[    8.187565] sw-ehci sw-ehci.2: USB 2.0 started, EHCI 1.00
[    8.227619] ehci_irq: port change detect
[    8.275965] hub 4-0:1.0: USB hub found
[    8.310025] hub 4-0:1.0: 1 port detected
[    8.360986] usbcore: registered new interface driver rtl8192cu
[    8.675912] usb 4-1: new high-speed USB device number 2 using sw-ehci
```

Also, I noticed that when I did get the dongles working, the internal realtek wifi which is on the USB was not listed in `lsusb`.  When the dongles aren't working I _sometimes_ have the realtek wifi listed in `lsusb`.  The wifi always appears on bus number 4 when it is there and when it isn't there there are no items listed on the 4th bus (including hubs).  The except to this is when I tried sunxi-v3.4.43.-r1 in which all of the devices were properly listed and there was even a 5th bus.

**Comment by tisdall** [comment]
I can't really get any further unless someone tells me how to get 8242195a7bb3d8ac3a829b8e7e0f226bd998d78a to compile

**Comment by tisdall** [comment]
@romanrm I tried modifying the script.bin in various ways with no success (using http://linux-sunxi.org/Fex_Guide#USB_control_flags as documentation).  I also tried https://github.com/linux-sunxi/sunxi-boards/blob/master/sys_config/a10/hackberry.fex with no success.

**Comment by arokux** [comment]
Which port are you using lower or upper? The lower (according to a wild
googling) should be USB OTG. Is this true?

Can you confirm that the upper port which is supposed to be a pure USB HOST
works with sunxi-v3.4.43-r1 and fails with sunxi-v3.4.43-r2?

**Comment by tisdall** [comment]
@arokux I'm not sure how to confirm the lower port is OTG, but that's what I found from Googling too.  However, I am trying both ports and neither have power.  I've also tried compiling the kernel as "host only" with the same results.

Yes, I can confirm that.  Both ports fail in sunxi-v3.4.43-r2 but work in r1.

**Comment by arokux** [comment]
On Mon, Nov 4, 2013 at 6:47 PM, tisdall notifications@github.com wrote:

> @arokux https://github.com/arokux I'm not sure how to confirm the lower
> port is OTG, but that's what I found from Googling too. However, I am
> trying both ports and neither have power. I've also tried compiling the
> kernel as "host only" with the same results.

As "host only" what do you mean by that?

Can you try to build sunxi-v3.4.43-r1/sunxi-v3.4.43-r2 with
CONFIG_USB_SW_SUNXI_USB removed from your .config and see what happens?
This will disable all the code for the USB OTG leaving you only with
support for the USB HOST.

In theory USB OTG and USB HOST should be independent...

**Comment by tisdall** [comment]
@arokux What I mean by "host only" is modifying the kernel options with `menuconfig` and selecting "host only" under "Device Drivers" -> "USB support" -> "SUNXI USB2.0 Dual Role Controller support" -> "USB0 Controller support"

I actually tried compiling with that option off already and got the same results (it was recommended in some post I read somewhere).  I just tried it again with the latest copy of "sunxi-3.4" and got the same results.

I believe I also tried with CONFIG_USB_GADGET off with no difference.

**Comment by tisdall** [comment]
@arokux after thinking a bit I realized you may be saying to shut that option off to get the 8242195a7bb3d8ac3a829b8e7e0f226bd998d78a to compile.  I checked out that hash and tried compiling with CONFIG_USB_SW_SUNXI_USB off and got the same problem.  So, 8242195a7bb3d8ac3a829b8e7e0f226bd998d78a is the check-in that created the problem.

**Comment by arokux** [comment]
On Mon, Nov 4, 2013 at 8:48 PM, tisdall notifications@github.com wrote:

> @arokux https://github.com/arokux after thinking a bit I realized you
> may be saying to shut that option off to get the 8242195https://github.com/linux-sunxi/linux-sunxi/commit/8242195a7bb3d8ac3a829b8e7e0f226bd998d78ato compile. I checked out that hash and tried compiling with
> CONFIG_USB_SW_SUNXI_USB off and got the same problem. So, 8242195https://github.com/linux-sunxi/linux-sunxi/commit/8242195a7bb3d8ac3a829b8e7e0f226bd998d78ais the check-in that created the problem.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/linux-sunxi/linux-sunxi/issues/158#issuecomment-27716527
> .

Let's do it hardcore. balage (IRC username) has found out the commit which
introduces the regression is

1d524c5e244b79a5a9570e028d2497fb17ffdfbb

All of the files that this commit touches reside in drivers/usb/sunxi_usb.

(You say it is  8242195a7bb3d8ac3a829b8e7e0f226bd998d78a but it also
touches things in drivers/usb/sunxi_usb only)

This directory contains the code for the USB OTG and should not interfere
with USB HOST in _no_ way. Let us make sure it is not built. Please remove
this line

https://github.com/linux-sunxi/linux-sunxi/blob/stage/sunxi-3.4/drivers/usb/Kconfig#L192

recompile the kernel and see if any of the USB ports works / has power
supplied to it.

Thanks for you hard work testing it, without you it won't be possible to
fix this bug since I do not own the hardware and cannot reproduce this
problem with my boards.

**Comment by tisdall** [comment]
@arokux  Sorry!  I did something stupid when compiling and missed recompiling the modules.  I can't get 8242195a7bb3d8ac3a829b8e7e0f226bd998d78a to compile and get the following:

```
$ make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf- -j3 INSTALL_MOD_PATH=output modules
   ...
ERROR: "sw_usb_disable_hcd" [drivers/net/wireless/rtl8192cu/8192cu.ko] undefined!
ERROR: "sw_usb_enable_hcd" [drivers/net/wireless/rtl8192cu/8192cu.ko] undefined!
```

I removed the module to complete the compile but still have no power on the USB ports.

So, I can confirm things are working on 95e472f5bf0fbbbc7f50969f81c9651bd642a94c but not on 8242195a7bb3d8ac3a829b8e7e0f226bd998d78a . (I retried both again to make sure)

Okay, that now brings us to your request...  I removed both `source "drivers/usb/sun3i_usb/Kconfig"` and `source "drivers/usb/sunxi_usb/Kconfig"` as they're both there for 8242195a7bb3d8ac3a829b8e7e0f226bd998d78a . (hopefully that's correct).   Result?  No change, the ports are still powerless.  Should we try removing `source "drivers/usb/otg/Kconfig"`?

**Comment by arokux** [comment]
On Mon, Nov 4, 2013 at 9:36 PM, tisdall notifications@github.com wrote:

> @arokux https://github.com/arokux Sorry! I did something stupid when
> compiling and missed recompiling the modules. I can't get 8242195https://github.com/linux-sunxi/linux-sunxi/commit/8242195a7bb3d8ac3a829b8e7e0f226bd998d78ato compile and get the following:
> 
> $ make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf- -j3 INSTALL_MOD_PATH=output modules
>    ...
> ERROR: "sw_usb_disable_hcd" [drivers/net/wireless/rtl8192cu/8192cu.ko] undefined!
> ERROR: "sw_usb_enable_hcd" [drivers/net/wireless/rtl8192cu/8192cu.ko] undefined!
> 
> I removed the module to complete the compile but still have no power on
> the USB ports.
> 
> So, I can confirm things are working on 95e472fhttps://github.com/linux-sunxi/linux-sunxi/commit/95e472f5bf0fbbbc7f50969f81c9651bd642a94cbut not on
> 8242195https://github.com/linux-sunxi/linux-sunxi/commit/8242195a7bb3d8ac3a829b8e7e0f226bd998d78a. (I retried both again to make sure)
> 
> Okay, that now brings us to your request... I removed both source
> "drivers/usb/sun3i_usb/Kconfig" and source "drivers/usb/sunxi_usb/Kconfig"as they're both there for
> 8242195https://github.com/linux-sunxi/linux-sunxi/commit/8242195a7bb3d8ac3a829b8e7e0f226bd998d78a. (hopefully that's correct).
> 
> sun3i is not relevant so it doesn't matter if it is there or not.
> 
> Result? No change, the ports are still powerless. Should we try removing source
> "drivers/usb/otg/Kconfig"?
> 
> ... Can you try if sunxi-v3.4.43-r1 works if

source "drivers/usb/sunxi_usb/Kconfig"

removed?

**Comment by techn** [comment]
@tisdall could you comment following #ifdef https://github.com/linux-sunxi/linux-sunxi/commit/8242195a7bb3d8ac3a829b8e7e0f226bd998d78a#diff-c5bd7effafe37f013728884246cb9af6R2345 ?

It is actually only change that is meaningful in that commit..

**Comment by tisdall** [comment]
@techn It looks like CONFIG_USB_PORT_POWER_MANAGEMENT is never mentioned anywhere else in the code, so shouldn't that always be false and not included?

**Comment by techn** [comment]
@tisdall Ok. I remembered wrong. CONFIG_USB_PORT_POWER_MANAGEMENT has effect only with later commit.

But in any case. It sounds impossiple that that commit has effect on USB1 or USB2, atleast when drivers/usb/sunxi_usb/Kconfig commented.

**Comment by tisdall** [comment]
@arokux okay, I tried 95e472f5bf0fbbbc7f50969f81c9651bd642a94c (which did work before) without sunxi_usb and I'm no longer able to see any of the USB devices on those ports.

**Comment by techn** [comment]
@tisdall could you try sunxi_usb directory removed with even older tag. (which doesnt have my usb commits)

**Comment by arokux** [comment]
On Mon, Nov 4, 2013 at 10:02 PM, Jari Helaakoski
notifications@github.comwrote:

> @tisdall https://github.com/tisdall could you try sunxi_usb directory
> removed with even older tag. (which doesnt have my usb commits)
> 
> —

Yes, as I said you should have tried sunxi-v3.4.43-r1 with

source "drivers/usb/sunxi_usb/Kconfig"

removed

Thank you for the hard work.

**Comment by tisdall** [comment]
@arukux that line doesn't exist in https://github.com/linux-sunxi/linux-sunxi/blob/sunxi-v3.4.43-r1/drivers/usb/Kconfig  I'll try with removing all of the sun?i_usb lines.

**Comment by arokux** [comment]
On Mon, Nov 4, 2013 at 10:12 PM, tisdall notifications@github.com wrote:

> @arukux that line doesn't exist in
> https://github.com/linux-sunxi/linux-sunxi/blob/sunxi-v3.4.43-r1/drivers/usb/KconfigI'll try with removing all of the sun?i_usb lines.

Yes, kill'em all!

**Comment by tisdall** [comment]
okay, killed them all and I don't have any power on the USB ports

**Comment by arokux** [comment]
On Mon, Nov 4, 2013 at 10:27 PM, tisdall notifications@github.com wrote:

> okay, killed them all and I don't have any power on the USB ports
> 
> Ok, we start to get grip on it!!! sun_i_usb/_ stuff interferes with USB
> HOST drivers.

@techn: what GPIOs are touched in sun*i_usb?

**Comment by techn** [comment]
@arokux I'm suspecting that hackberry's USB is somehow wrongly wired. Other devices doesn't have this problem.

Looks like OTG should be disabled with it. :/ 
Maybe latest code works when CONFIG_USB_SW_SUNXI_USB0_OTG is disabled (CONFIG_USB_SW_SUNXI_USB0_HOST_ONLY is set)

**Comment by arokux** [comment]
On Mon, Nov 4, 2013 at 10:35 PM, Jari Helaakoski
notifications@github.comwrote:

> @arokux https://github.com/arokux I'm suspecting that hackberry's USB
> is somehow wrongly wired. Other devices doesn't have this problem.
> 
> Looks like OTG should be disabled with it. :/
> Maybe latest code works when CONFIG_USB_SW_SUNXI_USB0_OTG is disabled
> (CONFIG_USB_SW_SUNXI_USB0_HOST_ONLY is set)

Maybe if we know what GPIOs are messed up we can work around?

**Comment by tisdall** [comment]
Not sure if this helps, but I remember seeing this in the script.bin when I was fiddling with it:  https://github.com/linux-sunxi/sunxi-boards/blob/master/sys_config/a10/hackberry.fex#L733

**Comment by arokux** [comment]
do you have some USB gadget with LED on it so you can detect if there is
power supplied to the USB port?

**Comment by tisdall** [comment]
@arokux The one bluetooth dongle I'm using has a light on it that flashes constantly once plugged in.  I don't see it light up at all when I'm not able to see the device on `lsusb`.

**Comment by tisdall** [comment]
@arokux oh.. better..  I have a portable battery that should light up when it's charging.  So, it should only rely on whether there's a current on the hub.

**Comment by techn** [comment]
@tisdall Or was OTG enabled on sunxi-v3.4.43-r1 ? Is USB_MANAGER enabled in working config?

**Comment by arokux** [comment]
Perfect.

Plug-in your battery into one of the usb ports (that was known to work) and
issue the following commands in U-Boot prompt. See for which of the
commands the LED will come ON on your battery.

gpio s PB09

gpio s PH06

gpio s PH12

**Comment by tisdall** [comment]
@techn `CONFIG_USB_SW_SUN4I_USB_MANAGER` is enabled.  I'm not sure how to test if the OTG capabilities are working.

**Comment by tisdall** [comment]
@arokux  What linaro/debian package do I need to install to get that `gpio` command?

**Comment by tisdall** [comment]
@arokux sorry.. missed the "u-boot prompt"..  I'm not sure if I can get access to that.  I have a USB-serial dongle, but I haven't been able to get it to work properly.

**Comment by arokux** [comment]
On Mon, Nov 4, 2013 at 10:57 PM, tisdall notifications@github.com wrote:

> @arokux https://github.com/arokux What linaro/debian package do I need
> to install to get that gpio command?

No packages, no kernel booting. You need to issue these commands in the
U-Boot prompt. (You can join #linux-sunxi on freenode and chat with me
there.)

**Comment by tisdall** [comment]
@arokux I'll look you up tomorrow morning...  I've got to go right now.

**Comment by arokux** [comment]
On Mon, Nov 4, 2013 at 11:01 PM, tisdall notifications@github.com wrote:

> @arokux https://github.com/arokux I'll look you up tomorrow morning...
> I've got to go right now.

Alright, thanks for all this hard work!

**Comment by tisdall** [comment]
@arokux okay, `gpio s PB09` managed to bring power to the USB port

**Comment by n1bs** [comment]
Hello Guys,

I have Hackberry and really hate that problem with USB powering, so will be glad to help with solving of that. I use latest kernel 3.4.67 and noticed that if I compile 8192cu as module, it will enable power on sw-echi2 during load. But without native module - it's always powered down. Also, during boot I can see that when SUNXI_USB is loaded - it's enabling power on USB for short time, and then almost immediately shutting it down. 

I can help with any kind of logs/debug/traces, and will be glad to fix some part of code - the problem for me is that I'm not experienced with kernel modules architecture and it takes too much of time for some simple stuff (locate proper entry point, and etc.). 

$~> dmesg | grep sw-ehci2
[    0.287352] [sw-ehci2]: probe, pdev->name: sw-ehci, pdev->id: 2, sw_ehci: 0xc0837cfc
[    0.287368] [sw-ehci2]: open clock
[    0.307379] [sw-ehci2]: Set USB Power ON
[    0.323158] [sw-ehci2]: probe, clock: SW_VA_CCM_AHBMOD_OFFSET(0x2414f), SW_VA_CCM_USBCLK_OFFSET(0x167); usb: 0x800(0x701), dram:(0x301, 0x301)
[    0.323173] [sw-ehci2]: disable ehci
[    0.323184] [sw-ehci2]: remove, pdev->name: sw-ehci, pdev->id: 2, sw_ehci: 0xc0837cfc
[    0.323994] [sw-ehci2]: Set USB Power OFF
[    0.324006] [sw-ehci2]: close clock

Dmesg once module loaded:

[   91.516173] [sw-ehci2]: enable ehci
[   91.516190] [sw-ehci2]: probe, pdev->name: sw-ehci, pdev->id: 2, sw_ehci: 0xc0837cfc
[   91.516213] [sw-ehci2]: open clock
[   91.536271] [sw-ehci2]: Set USB Power ON
[   91.557564] [sw-ehci2]: probe, clock: SW_VA_CCM_AHBMOD_OFFSET(0x2414f), SW_VA_CCM_USBCLK_OFFSET(0x167); usb: 0x800(0x701), dram:(0x301, 0x301)

$~> lsusb -t
/:  Bus 04.Port 1: Dev 1, Class=root_hub, Driver=sw-ehci/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Vendor Specific Class, Driver=rtl8192cu, 480M
/:  Bus 03.Port 1: Dev 1, Class=root_hub, Driver=sw-ohci/1p, 12M
/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=sw-ehci/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/4p, 480M
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=sw_hcd_host0/1p, 480M

Please let me know if there is IRC channel where we can discuss it to speedup the process. 

Best wishes!

**Comment by amery** [comment]
@n1bs #linux-sunxi on freenode

**Comment by techn** [comment]
@n1bs: we found that you can force power with CONFIG_USB_PORT_POWER_MANAGEMENT. It will be changed behind host only mode If that other patch gets approved.

Root cause for hackberry problems I suspect bad board design. OTG port should have sense pin grounded.
http://im.tech2.in.com/gallery/2012/jun/difference_circuit_271738111300.jpg

It would be nice to have this verified by manufacturer

**Comment by n1bs** [comment]
@techn Thanks a lot for fast reply. Trying to compile new kernel with defined CONFIG_USB_PORT_POWER_MANAGEMENT. 

Is that OTG port uses the same hub as built-in wifi module ? Checked with voltmeter - there is very short period during boot when power is supplied, and then disabled. 

Switching of mode to HOST_ONLY from OTG changes nothing :(

**Comment by n1bs** [comment]
Well, CONFIG_USB_PORT_POWER_MANAGEMENT helps with USB ports, so attached dongle is visible now. But unfortunately it's not turning ON built-in wifi module, so seems they use different ports. 

Logs after bootup:

```
$~> lsusb
Bus 001 Device 002: ID 0cf3:9271 Atheros Communications, Inc. AR9271 802.11n - attached to 1 of available USB port
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 003 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
Bus 002 Device 002: ID 05e3:0608 Genesys Logic, Inc. USB-2.0 4-Port HUB
Bus 002 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub

root| warder [Sat Nov 16 18:57:10] [ 295 ] ~ 
$~> lsusb -t
/:  Bus 03.Port 1: Dev 1, Class=root_hub, Driver=sw-ohci/1p, 12M
/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=sw-ehci/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/4p, 480M
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=sw_hcd_host0/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Vendor Specific Class, Driver=, 480M
```

dmesg after modprobe 8192cu:

```
[  178.730374] ERR: script_parser_fetch usb_controller_type failed
[  178.741869] [sw-ehci2]: enable ehci
[  178.744364] [sw-ehci2]: open clock
[  178.770721] [sw-ehci2]: Set USB Power ON
[  178.781365] sw-ehci sw-ehci.2: SW USB2.0 'Enhanced' Host Controller (EHCI) Driver
[  178.787559] sw-ehci sw-ehci.2: new USB bus registered, assigned bus number 4
[  178.791924] sw-ehci sw-ehci.2: irq 40, io mem 0x01c1c000
[  178.815416] sw-ehci sw-ehci.2: USB 2.0 started, EHCI 1.00
[  178.818048] ehci_irq: port change detect
[  178.832090] hub 4-0:1.0: USB hub found
[  178.838872] hub 4-0:1.0: 1 port detected
[  178.847434] usbcore: registered new interface driver rtl8192cu
[  179.156392] usb 4-1: new high-speed USB device number 2 using sw-ehci
[  179.591076] Error: Driver 'rtl8192cu' is already registered, aborting...

root| warder [Sat Nov 16 18:57:56] [ 299 ] ~ 
$~> lsusb
Bus 001 Device 002: ID 0cf3:9271 Atheros Communications, Inc. AR9271 802.11n
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 003 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
Bus 004 Device 002: ID 0bda:8176 Realtek Semiconductor Corp. RTL8188CUS 802.11n WLAN Adapter
Bus 004 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 002 Device 002: ID 05e3:0608 Genesys Logic, Inc. USB-2.0 4-Port HUB
Bus 002 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub

root| warder [Sat Nov 16 18:58:14] [ 300 ] ~ 
$~> lsusb -t
/:  Bus 04.Port 1: Dev 1, Class=root_hub, Driver=sw-ehci/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Vendor Specific Class, Driver=rtl8192cu, 480M
/:  Bus 03.Port 1: Dev 1, Class=root_hub, Driver=sw-ohci/1p, 12M
/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=sw-ehci/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/4p, 480M
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=sw_hcd_host0/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Vendor Specific Class, Driver=, 480M
```

So I want to know which file is responsible for this row:

[  178.770721] [sw-ehci2]: Set USB Power ON

May be it's possible just to force it to keep always ON.

**Comment by techn** [comment]
This could help:

```
usb:sunxi:Fix USB on Hackberry

Hackberry has OTG port wired as host-only port. It seems that wiring is done so that USB0-IDDET is done so that USB0-IDDET is not pulled down. This is wrong and leaves port without power.
This commit sets power on if port is defined as host-only in fex file or compile time.

This commit affects also auxtek-t003, mini-x_a10s, megafeis_a08 and mk802_a10s.


diff --git a/drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.c b/drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.c
index 846a01f..2399ec6 100644
--- a/drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.c
+++ b/drivers/usb/sunxi_usb/hcd/hcd0/sw_hcd0.c
@@ -1864,6 +1864,8 @@ static int sw_hcd_probe_host_only(struct platform_device *pdev)
        }

        usbc_no = 0;
+       hcd0_set_vbus_cnt--;
+       hcd0_enable--;

     memset(&g_sw_hcd_io, 0, sizeof(sw_hcd_io_t));
        ret = sw_hcd_io_init(usbc_no, pdev, &g_sw_hcd_io);
```

**Comment by tisdall** [comment]
@techn What needs to be done before it's committed into the repo?

**Comment by techn** [comment]
2013/12/6 tisdall notifications@github.com

**Comment by feuler** [comment]
The patch worked for me. My FTDI USB devices are recognized and working again.

But i get continous error messages now:

```
[ 2250.713228] [sw_hcd]: sw_hcd_urb_dequeue, sw_hcd(ef0958f0, 0x0, 0x3f),urb(eea8b500, 512, 0), dev = 5, ep = 1, dir = in
[ 2250.733236] sw_hcd_cleanup_urb: qh(0xee442c40,0x1,0x2), urb(0xeea8b500,512,0), ep(0xef095a00,4,0xee442c40,0x  (null))
[ 2250.748810] WRN:L2250(drivers/usb/sunxi_usb/hcd/core/sw_hcd_host.c):ERR: sw_hcd_host_rx, urb is NULL
[ 4752.294316] [sw_hcd]: sw_hcd_urb_dequeue, sw_hcd(ef0958f0, 0x0, 0x3f),urb(ee933380, 512, 0), dev = 4, ep = 1, dir = in
[ 4752.314324] sw_hcd_cleanup_urb: qh(0xee442000,0x1,0x2), urb(0xee933380,512,0), ep(0xef095a00,4,0xee442000,0x  (null))
[ 4752.329899] WRN:L2250(drivers/usb/sunxi_usb/hcd/core/sw_hcd_host.c):ERR: sw_hcd_host_rx, urb is NULL
[ 4929.365271] [sw_hcd]: sw_hcd_urb_dequeue, sw_hcd(ef0958f0, 0x0, 0x3f),urb(ee933a00, 512, 0), dev = 4, ep = 1, dir = in
[ 4929.385279] sw_hcd_cleanup_urb: qh(0xee4427c0,0x1,0x2), urb(0xee933a00,512,0), ep(0xef095a00,4,0xee4427c0,0x  (null))
[ 4929.400861] WRN:L2250(drivers/usb/sunxi_usb/hcd/core/sw_hcd_host.c):ERR: sw_hcd_host_rx, urb is NULL
[ 5294.817352] [sw_hcd]: sw_hcd_urb_dequeue, sw_hcd(ef0958f0, 0x0, 0x3f),urb(eebfb880, 512, 0), dev = 5, ep = 1, dir = in
```

**Comment by arokux** [comment]
Hi Jari,

can you maybe submit that patch as plain text inline? I think it got enough
testing already?

Thanks
Arokux

**Comment by techn** [comment]
arodux, Yes. I was about to send it to ML, but got problems with my virtual machine. And totally forgot whole thing.

Now patch is send to ML.
